pkg/logging/json_logger_test.go

```diff
@@ -0,0 +1,173 @@
+/*
+   Copyright The containerd Authors.
+
+   Licensed under the Apache License, Version 2.0 (the "License");
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*/
+
+package logging
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"os"
+	"path/filepath"
+	"testing"
+	"time"
+)
+
+func TestReadRotatedJSONLog(t *testing.T) {
+	tmpDir := t.TempDir()
+	file, err := os.CreateTemp(tmpDir, "logfile")
+	if err != nil {
+		t.Errorf("unable to create temp file, error: %s", err.Error())
+	}
+	stdoutBuf := &bytes.Buffer{}
+	stderrBuf := &bytes.Buffer{}
+	containerStopped := make(chan os.Signal)
+	// Start to follow the container's log.
+	fileName := file.Name()
+	go func() {
+		lvOpts := LogViewOptions{
+			Follow:  true,
+			LogPath: fileName,
+		}
+		viewLogsJSONFileDirect(lvOpts, file.Name(), stdoutBuf, stderrBuf, containerStopped)
+	}()
+
+	// log in stdout
+	expectedStdout := "line0\nline1\nline2\nline3\nline4\nline5\nline6\nline7\nline8\nline9\n"
+	dir := filepath.Dir(file.Name())
+	baseName := filepath.Base(file.Name())
+
+	// Write 10 lines to log file.
+	// Let ReadLogs start.
+	time.Sleep(50 * time.Millisecond)
+
+	type logContent struct {
+		Log    string `json:"log"`
+		Stream string `json:"stream"`
+		Time   string `json:"time"`
+	}
+
+	for line := 0; line < 10; line++ {
+		// Write the first three lines to log file
+		log := logContent{}
+		log.Log = fmt.Sprintf("line%d\n", line)
+		log.Stream = "stdout"
+		log.Time = time.Now().Format(time.RFC3339Nano)
+		time.Sleep(1 * time.Millisecond)
+		logData, _ := json.Marshal(log)
+		file.Write(logData)
+
+		if line == 5 {
+			file.Close()
+			// Pretend to rotate the log.
```

Comment:
small suggestion; currently in nerdctl we use  https://github.com/fahedouch/go-logrotate to rotate logs :)

Reply:
Rewriting using this library will be more complicated.